RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,453 @@
+    {
+      origins.emplace_back( default_origin,
+			    (m_operationMode == OperationMode::BEAM_SPOT_FIXED) ? m_zErrorBeamSpot : m_nSigmaZBeamSpot*bs->sigmaZ()
+			    );
+    }
+    else if (m_operationMode == OperationMode::VERTICES_FIXED || m_operationMode == OperationMode::VERTICES_SIGMA)
+    {
+      edm::Handle< reco::VertexCollection > vertices;
+      e.getByToken( token_vertex, vertices );
+      int n_vert = 0;
+      for ( const auto& v : (*vertices) )  
+      {
+	if ( v.isFake() || !v.isValid() ) continue;
+	origins.emplace_back( GlobalPoint( v.x(), v.y(), v.z() ),
+			      (m_operationMode == OperationMode::VERTICES_FIXED) ? m_zErrorVetex : m_nSigmaZVertex*v.zError()
+			      );
+	++n_vert;
+	if( n_vert >= m_maxNVertices ) break;
+      }
+      // no-vertex fall-back case:
+      if (origins.empty())
+      {
+	origins.emplace_back( default_origin,
+			      (m_nSigmaZBeamSpot > 0.) ? m_nSigmaZBeamSpot*bs->z0Error() : m_zErrorBeamSpot
+			      );
+      }
+    }
+    
+    const MeasurementTrackerEvent *measurementTracker = nullptr;
+    if(!token_measurementTracker.isUninitialized()) {
+      edm::Handle<MeasurementTrackerEvent> hmte;
+      e.getByToken(token_measurementTracker, hmte);
+      measurementTracker = hmte.product();
+    }
+
+    // create tracking regions (maximum MaxNRegions of them) in directions of the
+    // objects of interest (we expect that the collection was sorted in decreasing pt order)
+    int n_regions = 0;
+
+    if(m_seedingMode == SeedingMode::CANDIDATE_SEEDED) {
+
+      for(const auto& object : objs) {
+
+	GlobalVector direction( object.momentum().x(), object.momentum().y(), object.momentum().z() );	
+
+	for(const auto& origin : origins) {
+
+	  result.push_back(std::make_unique<RectangularEtaPhiTrackingRegion>(
+		               direction,			       
+			       origin.first,
+			       m_ptMin,
+			       m_originRadius,
+			       origin.second,
+			       m_deltaEta_Cand,
+			       m_deltaPhi_Cand,
+			       m_whereToUseMeasurementTracker,
+			       m_precise,
+			       measurementTracker,
+			       m_searchOpt
+			   ));
+	  ++n_regions;	  
+	  if( n_regions >= m_maxNRegions ) break;
+
+	}
+
+	if( n_regions >= m_maxNRegions ) break;
+
+      }
+
+    }
+     
+
+    else if(m_seedingMode == SeedingMode::POINT_SEEDED) {
+
+      for( const auto& direction : directionPoints ){
+
+	for(const auto& origin : origins) {
+	  	 
+	  result.push_back( std::make_unique<RectangularEtaPhiTrackingRegion>(
+			    direction, // GlobalVector
+			    origin.first, // GlobalPoint
+			    m_ptMin,
+			    m_originRadius,
+			    origin.second,
+			    m_deltaEta_Point,
+			    m_deltaPhi_Point,
+			    m_whereToUseMeasurementTracker,
+			    m_precise,
+			    measurementTracker,
+			    m_searchOpt
+			   ));
+	  ++n_regions;
+	  if( n_regions >= m_maxNRegions ) break;
+
+	}
+
+	if( n_regions >= m_maxNRegions ) break;
```

Comment:
Is this second check needed, given the one above inside the loop?

Reply:
Yes (as above).